File: optimal/src/solver.ts
import { kComb } from "./comb";
import { BitSet } from "./bitset";

export function greedyCover(
  n: number, k: number, j: number, s: number, minSGroups: number,
  pool: number[], deadline: number
): number[][] {
  const jSets = [...kComb(n, j)];
  const U = jSets.length;
  const kSets: number[][] = [];
  const bits: BitSet[] = [];
  
  for (const ks of kComb(n, k)) {
    if (Date.now() > deadline) break;         
    const b = new BitSet(U);
    
    if (s === j) {
      jSets.forEach((js, idx) => {
        let isSubset = true;
        for (const v of js) {
          if (!ks.includes(v)) {
            isSubset = false;
            break;
          }
        }
        if (isSubset) b.set(idx);
      });
    } else {
      jSets.forEach((js, idx) => {
        let c = 0;
        for (const v of js) if (ks.includes(v)) c++;
        if (c >= s) b.set(idx);
      });
    }
    
    kSets.push(ks); bits.push(b);
  }

  const coverCounts = new Array(U).fill(0);
  const sol: number[][] = [];
  let satisfied = false;
  
  while (!satisfied && sol.length < kSets.length) {
    let best = -1, gain = -1;
    let bestEfficiency = -1; 
    
    for (let i = 0; i < bits.length; i++) {
      let currentGain = 0;
      let uncoveredGain = 0;
      
      for (let j = 0; j < U; j++) {
        if (bits[i].get(j)) {
          if (coverCounts[j] < minSGroups) {
            currentGain++;
            if (coverCounts[j] === 0) uncoveredGain++;
          }
        }
      }
      
      const efficiency = uncoveredGain * 100 + currentGain;
      
      if (currentGain > 0 && (efficiency > bestEfficiency || 
         (efficiency === bestEfficiency && currentGain > gain))) {
        gain = currentGain;
        bestEfficiency = efficiency;
        best = i;
      }
    }
    
    if (best === -1 || gain === 0) break;

    sol.push(kSets[best].map(idx => pool[idx]));
    
    for (let j = 0; j < U; j++) {
      if (bits[best].get(j)) {
        coverCounts[j]++;
      }
    }
    
    satisfied = coverCounts.every(count => count >= minSGroups);
    
    bits.splice(best, 1);
    kSets.splice(best, 1);
  }

  if (!satisfied && s === j) {
    for (let u = 0; u < U && !satisfied; u++) {
      if (coverCounts[u] >= minSGroups) continue;

      let pick = -1;
      for (let i = 0; i < bits.length; i++) {
        if (bits[i].get(u)) { pick = i; break; }
      }
      if (pick === -1) break;        
  
      sol.push(kSets[pick].map(idx => pool[idx]));
      for (let jIdx = 0; jIdx < U; jIdx++) {
        if (bits[pick].get(jIdx)) coverCounts[jIdx]++;
      }
      satisfied = coverCounts.every(c => c >= minSGroups);
  
      bits.splice(pick, 1);
      kSets.splice(pick, 1);
    }
  }
  
  return sol;
}

export function localSearch(
  groups: number[][], pool: number[],
  n: number, j: number, s: number, minSGroups: number, deadline: number
): number[][] {
  const jSets = [...kComb(n, j)];
  
  const covered = (sol: number[][]) => {
    const coverCounts = new Array(jSets.length).fill(0);
    
    for (const kg of sol) {
      jSets.forEach((js, jIdx) => {
        if (s === j) {
          let isSubset = true;
          for (const idx of js) {
            if (!kg.includes(pool[idx])) {
              isSubset = false;
              break;
            }
          }
          if (isSubset) coverCounts[jIdx]++;
        } else {
          let c = 0;
          for (const idx of js) if (kg.includes(pool[idx])) c++;
          if (c >= s) coverCounts[jIdx]++;
        }
      });
    }
    
    return coverCounts.every(count => count >= minSGroups);
  };

  let improved = true;
  while (improved && Date.now() < deadline) {
    improved = false;
    
    for (let i = 0; i < groups.length; i++) {
      const test = groups.slice(0, i).concat(groups.slice(i + 1));
      if (covered(test)) { 
        groups = test; 
        improved = true; 
        break; }
    }
    
    
  }
  
  return groups;
}

export function simulatedAnnealing(
  initialGroups: number[][], pool: number[],
  n: number, j: number, s: number, k: number, minSGroups: number, 
  deadline: number
): number[][] {
  const jSets = [...kComb(n, j)];
  
  const covered = (sol: number[][]) => {
    const coverCounts = new Array(jSets.length).fill(0);
    
    for (const kg of sol) {
      jSets.forEach((js, jIdx) => {
        if (s === j) {
          let isSubset = true;
          for (const idx of js) {
            if (!kg.includes(pool[idx])) {
              isSubset = false;
              break;
            }
          }
          if (isSubset) coverCounts[jIdx]++;
        } else {
          let c = 0;
          for (const idx of js) if (kg.includes(pool[idx])) c++;
          if (c >= s) coverCounts[jIdx]++;
        }
      });
    }
    
    return coverCounts.every(count => count >= minSGroups);
  };
  
  const allGroups: number[][] = [];
  for (const ks of kComb(n, k)) {
    if (Date.now() > deadline - 3000) break;
    allGroups.push(ks.map(idx => pool[idx]));
  }
  
  let currentSolution = [...initialGroups];
  let bestSolution = [...initialGroups];
  
  const initialTemp = 15.0;
  const coolingRate = 0.96;
  const minTemp = 0.01;
  
  let temp = initialTemp;
  let iterations = 0;
  const MAX_ITERATIONS = 1500;
  
  while (temp > minTemp && iterations < MAX_ITERATIONS && Date.now() < deadline - 1000) {
    iterations++;
    
    let newSolution = [...currentSolution];
    const operation = Math.random();
    
    if (operation < 0.4 && newSolution.length > 0) {
      const replaceIndex = Math.floor(Math.random() * newSolution.length);
      const candidateGroups = allGroups.filter(g => 
        !newSolution.some(s => arraysEqual(s, g)));
      
      if (candidateGroups.length > 0) {
        const newGroup = candidateGroups[Math.floor(Math.random() * candidateGroups.length)];
        newSolution[replaceIndex] = newGroup;
      }
    } else if (operation < 0.7 && newSolution.length > 1) {
      const deleteIndex = Math.floor(Math.random() * newSolution.length);
      newSolution.splice(deleteIndex, 1);
    } else {
      const candidateGroups = allGroups.filter(g => 
        !newSolution.some(s => arraysEqual(s, g)));
      
      if (candidateGroups.length > 0) {
        const newGroup = candidateGroups[Math.floor(Math.random() * candidateGroups.length)];
        newSolution.push(newGroup);
      }
    }
    
    const isCovered = covered(newSolution);
    
    if (isCovered) {
      const deltaE = newSolution.length - currentSolution.length;
      
      if (deltaE < 0 || Math.random() < Math.exp(-deltaE / temp)) {
        currentSolution = newSolution;
        
        if (newSolution.length < bestSolution.length) {
          bestSolution = [...newSolution];
        }
      }
    }
    
    temp *= coolingRate;
  }
  
  return bestSolution;
}

function arraysEqual(a: number[], b: number[]): boolean {
  if (a.length !== b.length) return false;
  const sortedA = [...a].sort((x, y) => x - y);
  const sortedB = [...b].sort((x, y) => x - y);
  return sortedA.every((val, idx) => val === sortedB[idx]);
}

File: optimal/modules/optimal-module/android/src/main/java/expo/modules/optimalmodule/solver.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

class Solver {
    companion object {
        fun greedyCover(
            n: Int, k: Int, j: Int, s: Int, minSGroups: Int,
            pool: List<Int>, deadline: Long
        ): List<List<Int>> {
            val jSets = Comb.kComb(n, j)
            val U = jSets.size

            val kSetCount = Comb.calculateC(n, k)
            val kSets = ArrayList<List<Int>>(kSetCount)
            val bits = ArrayList<BitSet>(kSetCount)
            val mutex = Mutex()
            runBlocking {
                    val dispatcher = Dispatchers.Default
                    val jobs = Comb.kComb(n, k).map { ks ->
                        launch(dispatcher) {
                            if (System.currentTimeMillis() <= deadline) {
                                val b = BitSet(U)
                            if (s == j) {
                                jSets.forEachIndexed { idx, js ->
                                    var isSubset = true
                                    for (v in js) {
                                        if (!ks.contains(v)) {
                                            isSubset = false
                                            break
                                        }
                                    }
                                    if (isSubset) b.set(idx)
                                }
                            } else {
                                jSets.forEachIndexed { idx, js ->
                                    var c = 0
                                    for (v in js) if (ks.contains(v)) c++
                                    if (c >= s) b.set(idx)
                                }
                            }
                            mutex.withLock {
                            kSets.add(ks)
                            bits.add(b)
                            }
                        }}}
                jobs.forEach { it.join() }
            }
        
            val coverCounts = IntArray(U)
            val sol = mutableListOf<List<Int>>()
            var satisfied = false
            
            while (!satisfied && sol.size < kSets.size) {
                var best = -1
                var gain = -1
                var bestEfficiency = -1 
                
                for (i in bits.indices) {
                    var currentGain = 0
                    var uncoveredGain = 0 
                    
                    for (j in 0 until U) {
                        if (bits[i].get(j)) {
                            if (coverCounts[j] < minSGroups) {
                                currentGain++
                                if (coverCounts[j] == 0) uncoveredGain++
                            }
                        }
                    }
  
                    val efficiency = uncoveredGain * 100 + currentGain
                    
                    if (currentGain > 0 && (efficiency > bestEfficiency || 
                       (efficiency == bestEfficiency && currentGain > gain))) {
                        gain = currentGain
                        bestEfficiency = efficiency
                        best = i
                    }
                }
                
                if (best == -1 || gain == 0) break
                
                sol.add(kSets[best].map { idx -> pool[idx] })
                
                for (j in 0 until U) {
                    if (bits[best].get(j)) {
                        coverCounts[j]++
                    }
                }
                
                satisfied = coverCounts.all { count -> count >= minSGroups }
                
                bits.removeAt(best)
                kSets.removeAt(best)
            }

            if (!satisfied && s == j) {
                for (u in 0 until U) {
                    if (satisfied) break
                    if (coverCounts[u] >= minSGroups) continue
                
                    var pick = -1
                    for (i in bits.indices) {
                        if (bits[i].get(u)) { 
                            pick = i 
                            break 
                        }
                    }
                    if (pick == -1) break        
                
                    sol.add(kSets[pick].map { idx -> pool[idx] })
                    for (jIdx in 0 until U) {
                        if (bits[pick].get(jIdx)) coverCounts[jIdx]++
                    }
                    satisfied = coverCounts.all { c -> c >= minSGroups }
                
                    bits.removeAt(pick)
                    kSets.removeAt(pick)
                }
            }
            
            return sol
        }

        fun localSearch(
            groups: List<List<Int>>, pool: List<Int>,
            n: Int, j: Int, s: Int, minSGroups: Int, deadline: Long
        ): List<List<Int>> {
            val jSets = Comb.kComb(n, j)
            
            val covered = { sol: List<List<Int>> ->
                val coverCounts = IntArray(jSets.size) { 0 }
                
                for (kg in sol) {
                    jSets.forEachIndexed { jIdx, js ->
                        if (s == j) {
                            var isSubset = true
                            for (idx in js) {
                                if (!kg.contains(pool[idx])) {
                                    isSubset = false
                                    break
                                }
                            }
                            if (isSubset) coverCounts[jIdx]++
                        } else {
                            var c = 0
                            for (idx in js) if (kg.contains(pool[idx])) c++
                            if (c >= s) coverCounts[jIdx]++
                        }
                    }
                }
                
                coverCounts.all { count -> count >= minSGroups }
            }

            var improved = true
            var currentGroups = groups.toMutableList()
            
            while (improved && System.currentTimeMillis() < deadline) {
                improved = false
                
                for (i in currentGroups.indices) {
                    val test = currentGroups.toMutableList()
                    test.removeAt(i)
                    
                    if (covered(test)) { 
                        currentGroups = test
                        improved = true 
                        break 
                    }
                }
                
            }
            
            return currentGroups
        }

        fun simulatedAnnealing(
            initialGroups: List<List<Int>>, pool: List<Int>,
            n: Int, j: Int, s: Int, k: Int, minSGroups: Int, 
            deadline: Long
        ): List<List<Int>> {
            val jSets = Comb.kComb(n, j)
            
            val covered = { sol: List<List<Int>> ->
                val coverCounts = IntArray(jSets.size) { 0 }
                
                for (kg in sol) {
                    jSets.forEachIndexed { jIdx, js ->
                        if (s == j) {
                            var isSubset = true
                            for (idx in js) {
                                if (!kg.contains(pool[idx])) {
                                    isSubset = false
                                    break
                                }
                            }
                            if (isSubset) coverCounts[jIdx]++
                        } else {
                            var c = 0
                            for (idx in js) if (kg.contains(pool[idx])) c++
                            if (c >= s) coverCounts[jIdx]++
                        }
                    }
                }
                
                coverCounts.all { count -> count >= minSGroups }
            }
            
            val allGroups = mutableListOf<List<Int>>()
            for (ks in Comb.kComb(n, k)) {
                if (System.currentTimeMillis() > deadline - 3000) break 
                allGroups.add(ks.map { idx -> pool[idx] })
            }
            
            var currentSolution = initialGroups.toMutableList()
            var bestSolution = initialGroups.toMutableList()
            
            val initialTemp = 15.0
            val coolingRate = 0.96
            val minTemp = 0.01
            
            var temp = initialTemp
            var iterations = 0
            val MAX_ITERATIONS = 1500
            
            while (temp > minTemp && iterations < MAX_ITERATIONS && System.currentTimeMillis() < deadline - 1000) {
                iterations++

                val newSolution = currentSolution.toMutableList()
                val operation = Math.random()
                
                if (operation < 0.4 && newSolution.isNotEmpty()) {
                    val replaceIndex = (Math.random() * newSolution.size).toInt()
                    val candidateGroups = allGroups.filter { g -> 
                        !newSolution.any { s -> arraysEqual(s, g) }
                    }
                    
                    if (candidateGroups.isNotEmpty()) {
                        val newGroup = candidateGroups[(Math.random() * candidateGroups.size).toInt()]
                        newSolution[replaceIndex] = newGroup
                    }
                } else if (operation < 0.7 && newSolution.size > 1) {
                    val deleteIndex = (Math.random() * newSolution.size).toInt()
                    newSolution.removeAt(deleteIndex)
                } else {
                    val candidateGroups = allGroups.filter { g -> 
                        !newSolution.any { s -> arraysEqual(s, g) }
                    }
                    
                    if (candidateGroups.isNotEmpty()) {
                        val newGroup = candidateGroups[(Math.random() * candidateGroups.size).toInt()]
                        newSolution.add(newGroup)
                    }
                }
                
                val isCovered = covered(newSolution)
                
                if (isCovered) {
                    val deltaE = newSolution.size - currentSolution.size
                    
                    if (deltaE < 0 || Math.random() < Math.exp(-deltaE / temp)) {
                        currentSolution = newSolution
                        
                        if (newSolution.size < bestSolution.size) {
                            bestSolution = newSolution.toMutableList()
                        }
                    }
                }
                
                temp *= coolingRate
            }
            
            return bestSolution
        }

        private fun arraysEqual(a: List<Int>, b: List<Int>): Boolean {
            if (a.size != b.size) return false
            val sortedA = a.sorted()
            val sortedB = b.sorted()
            return sortedA.zip(sortedB).all { (x, y) -> x == y }
        }
    }
}
